include/rocksdb/iterator.h

```diff
@@ -123,14 +123,36 @@ class Iterator : public Cleanable {
    * checking ReadOptions::iterate_lower_bound,
    * false otherwise.
    */
-  virtual bool has_lower_bound() const { return false; }
+  virtual bool check_lower_bound() const { return false; }
```

Comment:
I am a little confused as to when/how these methods will be used.  They are new and the only Iterators that implement them are the test ones (even the BaseDelta does not implement it).  It seems like we might need a solution without these new methods, as all non-test solutions would effectively be without them.  Or am I missing something?

Reply:
I can certainly remove them. They are there to allow an optimisation where when you have an iterator wrapping an iterator (several places in Rocks) if your underlying iterator handles upper and lower bounds, you don't need to check them again.

I think it would be better to keep them, and expand the use of them. For example, `ForwardIterator` and `BackwardIterator` should likely implement them as they already handle upper and lower bounds checks respectively.

Reply:
Should the BaseDeltaIterator implement these functions?  

Not to complicate your world (and I am in no way suggesting you fix this, perhaps document it) but if the BaseIterator had bounds on it but the ReadOptions did not, wouldn't you potentially get really strange results?  That sounds more like something that should be an assert somewhere...

Reply:
The ReadOptions is passed to the BaseIterator - so I am not quite sure how you mean that one could have bounds but not the other?

Reply:
That is only true in the test code.  Outside of tests ,the "NewIteratorFromBase" takes in an Iterator, which becomes the base iterator.  In this case, the base iterator could have very different bounds than the DeltaIterator.

Or am I missing something?

I am not suggesting that should be legal BTW, only if it happens I do not know what the results would look like.

Reply:
Hmmm I will add some tests for that too...

Reply:
So this is already handled, and I just added the tests: `TestIteraratorWithBaseUpperBoundOnBaseNoUpperBoundOnBatch` `TestIteraratorWithBaseNoUpperBoundOnBaseUpperBoundOnBatch` for further evidence of this :-)